docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,240 @@
+endpoint for the device will get the node device name passed in by
+virt-handler. Virt-handler, which has to invoke the plugin from the host
+network namespace view, will fetch the node IP from the `Node` object in the
+cluster, and derive the device from it.
+
+The section on the `Node` from which the IP needs to be taken:
+
+```yaml
+status:
+  addresses:
+  - address: 192.168.121.180
+    type: InternalIP
+```
+
+Based on this IP, the master device can be derived and passed in to the plugin.
+It should be sufficient to derive this information once, on virt-handler start.
+
+When a VM contains a network configuration like this:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: NodeNetwork
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+```
+
+The plugin get's passed in the following config:
+
+```json
+{
+    "name": "myvm-nodenetwork",
+    "type": "macvlan",
+    "master": "eth0",
+    "mode": "bridge",
+    "ipam": {}
+}
+```
+
+which will do the mapping. No IPAM module is specified, since the VM can use
+plain DHCP to get a host network IP.
+
+Finally, to forbid unautorized users to connect the host networks, we can think
+about a similar concept like a `PodSecurityPolicy`, which can be tied to RBAC
+access.
+
+## macvlan/macvtap
+
+```
+  Host               Container                               VM
+| ---------------- | ------------------------------------- | ------------ |
+|           KubeVirt CNI          |                     Libvirt           |
+| ------------------------------- | --------------------------------------| 
+
+  eth0 ----------------macvlan1 -------> macvtap1 -------------> /dev/ethX
+
+```
+
+virt-handler calls a KubeVirt CNI plugin. The plugin creates a veth interface,
+moves one end into the container, and connects the other end with the host
+interface. The plugin is also responsible for generating/providing MAC
+addresses, if none are provided. The plugin returns the name of the veth
+endpoint in the container, and the MAC address. Libvirt then uses it's
+`direct` attach capability, by creating a macvtap interface, and configures it
+with the MAC address provided.
+
+To check: 
+
+ * Does libvirt need udev in the container or does it the work itself?
+
+Advantage:
+
+ * At the end macvtaps are used, which means that the VM get's the mac address
+   assigned by libvirt
+
+Disadvantage:
+
+ * The VM can't directly communicate with the underlying node, if switches are
+   not configured to mirror traffic.
+
+Based on this proposed method, at the end, one would see
+
+
+```xml
+<interface type='direct'>
+  <source dev='macvlan1' mode='bridge'/>
+</interface>
+```
+
+in the final domain XML.
+
+### How can the node and the nested node communicate with each other?
+
+In case of `macvlan` and `macvtap`, normally hosts and guests/containers can't
+communicate with each other. However, we can add an additional `macvlan`
```

Comment:
@rmohr I missed the part where you modify the routes.

It's possible this solution could work.  Mutating the routing table on the host still has me a little concerned. The unknown for me here is whether or not the routing table change would conflict with any of the pod networking solutions (weave, calico, ovs, etc...)

I'm not sure anyone can definitively answer that for us. At this point I think we just need a POC that we can drop into different scenarios and see for ourselves.